File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/hr/Department.java
package org.babyfish.jimmer.sql.model.hr;

import org.babyfish.jimmer.Formula;
import org.babyfish.jimmer.jackson.JsonConverter;
import org.babyfish.jimmer.jackson.LongToStringConverter;
import org.babyfish.jimmer.sql.*;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@KeyUniqueConstraint(noMoreUniqueConstraints = true)
public interface Department {

    @Id
    @JsonConverter(LongToStringConverter.class)
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key
    String name();

    @LogicalDeleted
    long deletedMillis();

    @OneToMany(mappedBy = "department")
    List<Employee> employees();

    @IdView("employees")
    List<Long> employeeIds();

    @Formula(sql = "(select count(*) from employee where department_id = %alias.id)")
    long employeeCount();
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/hr/Department.kt
package org.babyfish.jimmer.sql.kt.model.hr

import com.fasterxml.jackson.annotation.JsonFormat
import org.babyfish.jimmer.Formula
import org.babyfish.jimmer.jackson.JsonConverter
import org.babyfish.jimmer.jackson.LongToStringConverter
import org.babyfish.jimmer.sql.*
import testpkg.annotations.ConsiderAs
import testpkg.annotations.Module
import testpkg.annotations.Type
import java.time.LocalDateTime

@Entity
@Module("hr")
interface Department {

    @Id
    @get:ConsiderAs(types = [Type(String::class)])
    @JsonConverter(LongToStringConverter::class)
    val id: Long

    @Key
    @JsonConverter(ConverterForIssue937::class)
    val name: String

    @get:JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    @LogicalDeleted("now")
    val deletedTime: LocalDateTime?

    @OneToMany(mappedBy = "department")
    val employees: List<Employee>

    @IdView("employees")
    val employeeIds: List<Long>

    @Formula(sql = "(select count(*) from employee where department_id = %alias.id)")
    val employeeCount: Long

    @Formula(dependencies = ["id", "name"])
    val description: String
        get() = "$id-${name.uppercase()}"
}